### PR TITLE
3D Tiles binary functions

### PR DESCRIPTION
**Pull request by leesafini**
For #4643 

This PR supports four binary functions via a data-driven approach:
- `atan2`
- `pow`
- `min`
- `max`

### CODE REVIEW

**Comment by leesafini** [comment]
Here are some screenshots:

image
image
image
image

File: Apps/Sandcastle/gallery/3D Tiles Point Cloud Styling.html [hunk 1]
```diff
@@ -118,9 +118,29 @@
     color : "color() * acos(degrees(${temperature})) + color() * asin(${temperature}) + color() * atan(${temperature})"
 });
 
+addStyle('ATan2', {
+    color : "color() * atan2(${POSITION}[0],${temperature}) + color() * atan2(${POSITION}[1],${temperature}) + color() * atan2(${POSITION}[2],${temperature})",
+    pointSize : "5"
+});
```
Comment (lilleyse) [inline Apps/Sandcastle/gallery/3D Tiles Point Cloud Styling.html]: With https://github.com/AnalyticalGraphicsInc/cesium/pull/4628 in just incorporate atan2 to the Trigonometric Functions style.

File: Apps/Sandcastle/gallery/3D Tiles Point Cloud Styling.html [hunk 2]
```diff
@@ -118,9 +118,29 @@
...
+
+addStyle('Min', {
+    color : "color() * min(${POSITION}[0],0.75)",
+    pointSize : "5"
+});
+
+addStyle('Max', {
+    color : "color() * max(${POSITION}[2],0.25)",
+    pointSize : "5"
```
Comment (lilleyse) [inline Apps/Sandcastle/gallery/3D Tiles Point Cloud Styling.html]: Min and max could be combined into one style, min could affect the red channel and max could affect the green channel.

File: Specs/Scene/ExpressionSpec.js [hunk 3]
```diff
@@ -988,6 +988,78 @@ defineSuite([
         }).toThrowDeveloperError();
     });
 
+    it('evaluates atan2 function', function() {
+        var expression = new Expression('atan2(0,1)');
+        expect(expression.evaluate(frameState, undefined)).toEqual(0.0);
+
+        var expression = new Expression('atan2(1,0)');
+        expect(expression.evaluate(frameState, undefined)).toEqual(0.5*PI);
```
Comment (lilleyse) [inline Specs/Scene/ExpressionSpec.js]: If this test fails you may need to use `toEqualEpsilon` like here: https://github.com/AnalyticalGraphicsInc/cesium/pull/4628#issuecomment-261650132.

`PI` should be `Math.PI`

File: Specs/Scene/ExpressionSpec.js [hunk 4]
```diff
@@ -1807,6 +1879,34 @@ defineSuite([
...
+        var expression = new Expression('min(3,5)');
+        var shaderExpression = expression.getShaderExpression('', {});
+        var expected = 'min(3,5)';
+        expect(shaderExpression).toEqual(expected);
+    });
+
+    it('gets shader expression for max', function() {
+        var expression = new Expression('max(3,5)');
+        var shaderExpression = expression.getShaderExpression('', {});
+        var expected = 'max(3,5)';
```
Comment (lilleyse) [inline Specs/Scene/ExpressionSpec.js]: All of these expected lines need to be like `max(3.0, 5.0)`. This came up in @Dylan-Brown's PR too: https://github.com/AnalyticalGraphicsInc/cesium/pull/4578#issuecomment-258256974

**Comment by lilleyse** [comment]
Also make sure you have merged the latest 3d-tiles into this branch. The `frameState` is now required by all the evaluate functions.

File: Source/Scene/Expression.js [hunk 5]
```diff
@@ -45,6 +45,13 @@ define([
         }
     };
 
+    var binaryFunctions = {
+        atan2 : Math.atan,
```
Comment (lilleyse) [inline Source/Scene/Expression.js]: This should still be `Math.atan2`. Only the glsl side needs to change to `atan`.

File: Source/Scene/Expression.js [hunk 6]
```diff
@@ -1182,6 +1191,10 @@ define([
                     return '(' + left + ' == ' + right + ')';
                 } else if (value === '!==') {
                     return '(' + left + ' != ' + right + ')';
+                } else if (value === 'atan2') {
+                    return 'atan' + '(' + left + ', ' + right + ')';
```
Comment (lilleyse) [inline Source/Scene/Expression.js]: Can be slightly condensed to `'atan(' + `

**Comment by leesafini** [comment]
@lilleyse Updated, but integration check is still failing. Not sure what the problem is.

**Comment by lilleyse** [comment]
errors

It looks like you just need to add spaces after the commas in the expected strings.

**Comment by lilleyse** [comment]
And `min(3,5)` should be `min(3.0, 5.0)`.

**Comment by leesafini** [comment]
@lilleyse Updated, still failing. I checked the log and I don't think it's one of my tests that's failing.

**Comment by lilleyse** [comment]
There are jsHint errors.

**Comment by leesafini** [comment]
Thanks, I see it now. Is there anyway to tell exactly where in Expression and ExpressionSpec things are failing?

**Comment by lilleyse** [comment]
There should be line numbers corresponding with the errors. Also if you run jsHint locally it should tell you as well.

**Comment by leesafini** [comment]
@lilleyse Think I fixed it!

File: Apps/Sandcastle/gallery/3D Tiles Point Cloud Styling.html [hunk 7]
```diff
@@ -115,12 +115,23 @@
...
+});
+
+addStyle('Pow', {
+    color : "color() * pow(${temperature}, 3)",
+    pointSize : "5"
+});
+
+addStyle('Min and Max', {
+    color : "rgb(min(${POSITION}[0],0.75)*255, max(${POSITION}[2],0.25)*255, 255",
```
Comment (lilleyse) [inline Apps/Sandcastle/gallery/3D Tiles Point Cloud Styling.html]: Close the rgb with `)`.

Also add spaces after commas and between `*` so its easier to read.

**Comment by lilleyse** [comment]
Looks good @leesafini!

**Pull request merged**